cmdstanpy/stanfit.py

```diff
@@ -727,47 +727,43 @@ def draws_as_dataframe(
+    def stan_variable(self, name: str) -> pd.DataFrame:
         """
-        Return a new ndarray which contains the set of post-warmup draws
+        Return a new DataFrame which contains the set of post-warmup draws
         for the named Stan program variable.  Flattens the chains.
         Underlyingly draws are in chain order, i.e., for a sample
         consisting of N chains of M draws each, the first M array
         elements are from chain 1, the next M are from chain 2,
         and the last M elements are from chain N.
 
-        * If the variable is a scalar variable, this returns a 1-d array,
-          length(draws X chains).
-        * If the variable is a vector, this is a 2-d array,
+        * If the variable is a scalar variable, the shape of the DataFrame is
+          ( draws X chains, 1).
+        * If the variable is a vector, the shape of the DataFrame is
           shape ( draws X chains, len(vector))
-        * If the variable is a matrix, this is a 3-d array,
-          shape ( draws X chains, matrix nrows, matrix ncols ).
-        * If the variable is an array with N dimensions, this is an N+1-d array,
-          shape ( draws X chains, size(dim 1), ... size(dim N)).
+        * If the variable is a matrix, the shape of the DataFrame is
+          shape ( draws X chains, size(dim 1) X size(dim 2) )
+        * If the variable is an array with N dimensions, the shape of the DataFrame is
+          shape ( draws X chains, size(dim 1) X ... X size(dim N)).
 
         :param name: variable name
         """
         if name not in self._stan_variable_dims:
             raise ValueError('unknown name: {}'.format(name))
         self._assemble_draws()
         dim0 = self.num_draws * self.runset.chains
-        dims = self._stan_variable_dims[name]
-        if dims == 1:
-            idx = self.column_names.index(name)
-            return self._draws[self._draws_warmup :, :, idx].reshape(
-                (dim0,), order='A'
-            )
-        else:
-            idxs = [
-                x[0]
-                for x in enumerate(self.column_names)
-                if x[1].startswith(name + '.')
-            ]
-            var_dims = [dim0]
-            var_dims.extend(dims)
-            return self._draws[
-                self._draws_warmup :, :, idxs[0] : idxs[-1] + 1
-            ].reshape(tuple(var_dims), order='A')
+        dims = np.prod(self._stan_variable_dims[name])
+        pattern = r'^{}(\.\d+)*$'.format(name)
+        names, idxs = [], []
+        for i, column_name in enumerate(self.column_names):
+            if re.search(pattern, column_name):
+                names.append(column_name)
+                idxs.append(i)
+        return pd.DataFrame(
+            self._draws[
+                self._draws_warmup:, :, idxs
+            ].reshape((dim0, dims), order='A'),
+            columns=names
+        )
 
     def stan_variables(self) -> Dict:
         """
```

Comment:
What should we do for this function? Or maybe we could basically copy the `stan_variable`, but just change it so it works with multiple names. I don't mind of duplicated code, but I would like to create dataframe only once.

Would it make sense to use concat? And suggest users to learn filter

https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.filter.html

cc @mitzimorris

Reply:
It could return the whole DataFrame as default and a DataFrame containing only the requested variables if names are given. But I guess that would make the stan_variable function kind of redundant.

Reply:
True, we could make it so that stab_variable would call stan_variables.

What is the output in CmdStanR?

Reply:
I don't know because I don't use R but I'll have a look at it later. Does stan_variables have to be a function? How about a cached property and stan_variable just uses filter on the columns?

Reply:
Just had a look at the CmdStanR and the equivalent seems to be CmdStanMCMC::draws() which returns a iterations X chains X variables array or a iterations X chains array when called with a variable name as argument.

Reply:
maybe this obviates the need for `stan_variables`?

Reply:
what's useful about `stan_variables` and corresponding `stan_variable_dims` is that the returned dict gives you the names of all Stan variables present in the output - we need something that combines this and `stan_variable_dims`, information-wise.

Reply:
So maybe we could still return a dict of dataframes in `stan_variables`, that sounds fine by me.